presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetCompressionUtils.java

```diff
@@ -0,0 +1,94 @@
+import io.airlift.slice.Slice;
+import parquet.hadoop.metadata.CompressionCodecName;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.zip.GZIPInputStream;
+
+import static io.airlift.slice.Slices.EMPTY_SLICE;
+import static io.airlift.slice.Slices.wrappedBuffer;
+import static java.util.Objects.requireNonNull;
+import static parquet.hadoop.metadata.CompressionCodecName.GZIP;
+import static parquet.hadoop.metadata.CompressionCodecName.SNAPPY;
+import static parquet.hadoop.metadata.CompressionCodecName.UNCOMPRESSED;
+import static sun.misc.Unsafe.ARRAY_BYTE_BASE_OFFSET;
+
+public class ParquetCompressionUtils
+{
+    private ParquetCompressionUtils() {}
+
+    public static Slice decompress(CompressionCodecName codec, Slice input, int uncompressedSize)
+            throws IOException
+    {
+        requireNonNull(input, "input is null");
+
+        if (input.length() == 0) {
+            return EMPTY_SLICE;
+        }
+
+        switch (codec) {
+            case GZIP:
+                return decompressGzip(input, uncompressedSize);
+            case SNAPPY:
+                return decompressSnappy(input, uncompressedSize);
+            case UNCOMPRESSED:
+                return input;
+            default:
+                // TODO add LZO support using pure java aircompressor
+                throw new ParquetCorruptionException("Codec not supported in Parquet: " + codec);
+        }
+    }
+
+    private static Slice decompressSnappy(Slice input, int uncompressedSize)
+    {
+        byte[] inArray = (byte[]) input.getBase();
+        int inOffset = (int) (input.getAddress() - ARRAY_BYTE_BASE_OFFSET);
+        int inLength = input.length();
+
+        byte[] buffer = new byte[uncompressedSize];
+        SnappyDecompressor decompressor = new SnappyDecompressor();
+        decompressor.decompress(inArray, inOffset, inLength, buffer, 0, uncompressedSize);
+        return wrappedBuffer(buffer);
+    }
+
+    private static Slice decompressGzip(Slice input, int uncompressedSize)
+            throws IOException
+    {
+        // TODO use pure java aircompressor
+        ByteArrayOutputStream uncompressedStream = new ByteArrayOutputStream();
+        byte[] buffer = new byte[uncompressedSize];
+        InputStream gzipInputStream = null;
+        try {
+            gzipInputStream = new GZIPInputStream(new ByteArrayInputStream(input.getBytes()));
+            int bytesRead;
+            while ((bytesRead = gzipInputStream.read(buffer)) != -1) {
+                uncompressedStream.write(buffer, 0, bytesRead);
+            }
+            return wrappedBuffer(uncompressedStream.toByteArray());
+        }
+        catch (IOException e) {
+            throw new ParquetCorruptionException("Exception decoding gzip Parquet file: " + e.getMessage());
+        }
+        finally {
+            gzipInputStream.close();
```

Comment:
This can cause an NPE if the code above throws an exception during construction of the GZIPInputStream.